File: test_app/integration_tests/android_mocks/src/main/java/com/voxeet/asserts/FilePresentationServiceAsserts.java
package com.voxeet.asserts;

import android.net.Uri;
import android.util.Pair;

import com.voxeet.VoxeetSDK;
import com.voxeet.sdk.models.v1.FilePresentationConverted;

import java.io.File;
import java.net.MalformedURLException;
import java.util.Map;

import io.dolby.asserts.AssertUtils;
import io.dolby.asserts.MethodDelegate;

public class FilePresentationServiceAsserts implements MethodDelegate {
    @Override
    public void onAction(String methodName, Map<String, Object> args, Result result) {
        try {
            switch (methodName) {
                case "assertConvertArgs":
                    assertConvertArgs(args);
                    break;
                case "assertSetPageArgs":
                    assertSetPageArgs(args);
                    break;
                case "assertStartArgs":
                    assertStartArgs(args);
                    break;
                case "assertStopArgs":
                    assertStopArgs(args);
                    break;
                case "setGetImageReturn":
                    setGetImageReturn(args);
                    break;
                case "assertGetImageArgs":
                    assertGetImageArgs(args);
                    break;
                case "setGetThumbnailReturn":
                    setGetThumbnailReturn(args);
                    break;
                case "assertGetThumbnailArgs":
                    assertGetThumbnailArgs(args);
                    break;
                default:
                    result.error(new NoSuchMethodError("Method: " + methodName + " not found in " + getName() + " method channel"));
                    return;
            }
            result.success();
        } catch (AssertionFailed exception) {
            result.failed(exception);
        } catch (Exception ex) {
            result.error( ex);
        }
    }

    private void assertGetThumbnailArgs(Map<String, Object> args) throws AssertionFailed {
        boolean mockHasRun = VoxeetSDK.filePresentation().thumbnailHasRun;
        if(args.containsKey("hasRun")) {
            AssertUtils.compareWithExpectedValue(mockHasRun, args.get("hasRun"), "HasRun is incorrect");
        }

        Pair<String, Integer> mockArgs = VoxeetSDK.filePresentation().thumbnailArgs;
        if(args.containsKey("page")) {
            AssertUtils.compareWithExpectedValue(mockArgs.second, args.get("page"), "Page is incorrect");
        }
    }

    private void setGetThumbnailReturn(Map<String, Object> args) throws MalformedURLException {
        if(args.containsKey("url")) {
            VoxeetSDK.filePresentation().thumbnailReturn = new java.net.URL(args.get("url").toString());
        }
    }

    private void assertGetImageArgs(Map<String, Object> args) throws AssertionFailed {
        boolean mockHasRun = VoxeetSDK.filePresentation().imageHasRun;
        if(args.containsKey("hasRun")) {
            AssertUtils.compareWithExpectedValue(mockHasRun, args.get("hasRun"), "HasRun is incorrect");
        }

        Pair<String, Integer> mockArgs = VoxeetSDK.filePresentation().imageArgs;
        if(args.containsKey("page")) {
            AssertUtils.compareWithExpectedValue(mockArgs.second, args.get("page"), "Page is incorrect");
        }
    }

    private void setGetImageReturn(Map<String, Object> args) throws MalformedURLException {
        if(args.containsKey("url")) {
            VoxeetSDK.filePresentation().imageReturn = new java.net.URL(args.get("url").toString());
        }
    }

    private void assertStopArgs(Map<String, Object> args) throws AssertionFailed {
        boolean mockHasRun = VoxeetSDK.filePresentation().stopHasRun;
        if(args.containsKey("hasRun")) {
            AssertUtils.compareWithExpectedValue(mockHasRun, args.get("hasRun"), "HasRun is incorrect");
        }
    }

    private void assertStartArgs(Map<String, Object> args) throws AssertionFailed {
        boolean mockHasRun = VoxeetSDK.filePresentation().startHasRun;
        if(args.containsKey("hasRun")) {
            AssertUtils.compareWithExpectedValue(mockHasRun, args.get("hasRun"), "HasRun is incorrect");
        }

        FilePresentationConverted mockArgs = VoxeetSDK.filePresentation().startArgs;
        if(args.containsKey("id")) {
            AssertUtils.compareWithExpectedValue(mockArgs.fileId, args.get("id").toString(), "Id is incorrect");
        }

        if(args.containsKey("imageCount")) {
            AssertUtils.compareWithExpectedValue(mockArgs.nbImageConverted, AssertUtils.getInteger(args.get("imageCount")), "Image count is incorrect");
        }
    }

    private void assertSetPageArgs(Map<String, Object> args) throws AssertionFailed {
        boolean mockHasRun = VoxeetSDK.filePresentation().updateHasRun;
        if(args.containsKey("hasRun")) {
            AssertUtils.compareWithExpectedValue(mockHasRun, args.get("hasRun"), "HasRun is incorrect");
        }

        Pair<String, Integer> mockArgs = VoxeetSDK.filePresentation().updateArgs;
        if(args.containsKey("page")) {
            AssertUtils.compareWithExpectedValue(mockArgs.second, args.get("page"), "Page is incorrect");
        }
    }

    @Override
    public String getName() {
        return "IntegrationTesting.FilePresentationServiceAsserts";
    }

    private void assertConvertArgs(Map<String, Object> args) throws AssertionFailed {
        boolean mockHasRun = VoxeetSDK.filePresentation().convertHasRun;
        if (args.containsKey("hasRun")) {
            AssertUtils.compareWithExpectedValue(mockHasRun, args.get("hasRun"), "HasRun is incorrect");
        }

        File mockArgs = VoxeetSDK.filePresentation().convertArgs;
        if (args.containsKey("uri")) {
            Uri authFile = Uri.parse(mockArgs.getPath());
            android.util.Log.d("[KB]", "file schema: " + authFile.getScheme());
            android.util.Log.d("[KB]", "file path: " + authFile.getPath());
            String fileName = authFile.getScheme() + "://" + authFile.getPath();
            AssertUtils.compareWithExpectedValue(fileName, (String) args.get("uri"), "Uri doesn't match");
        }
    }
}

File: test_app/integration_tests/ios_mocks/VoxeetSDKMock/Classes/FilePresentationServiceAsserts.swift
import Foundation

public class FilePresentationServiceAsserts {
    
    private static var instance: FilePresentationServiceAsserts?
    public static func create() {
        instance = .init()
    }
    
    var methodChannel: AssertsMethodChannel?
    
    public init() {
        assertsMethodChannelFactory.createAssertMethodChannel(
            forSdkAsserts: self,
            channelName: "IntegrationTesting.FilePresentationServiceAsserts"
        )
    }
    
    private func assertGetCurrentArgs(args: [String: Any]) throws {
        let mockArgs = VoxeetSDK.shared.filePresentation.current
        try ifKeyExists(arg: args, key: "id") { (id: String) in
            try nativeAssertEquals(mockArgs?.id, id)
        }
        
        try ifKeyExists(arg: args, key: "ownerId") { (ownerId: String) in
            try nativeAssertEquals(mockArgs?.owner.id, ownerId)
        }
        
        try ifKeyExists(arg: args, key: "imageCount") { (imageCount: Int) in
            try nativeAssertEquals(mockArgs?.imageCount, imageCount)
        }
        
        try ifKeyExists(arg: args, key: "position") { (position: Int) in
            try nativeAssertEquals(mockArgs?.position, position)
        }
    }
    
    private func assertConvertArgs(args: [String: Any]) throws {
        let mockHasRun = VoxeetSDK.shared.filePresentation.convertHasRun
        try ifKeyExists(arg: args, key: "hasRun") { (hasRun: Bool) in
            try nativeAssertEquals(mockHasRun, hasRun)
        }
        
        let mockArgs = VoxeetSDK.shared.filePresentation.convertArgs
        try ifKeyExists(arg: args, key: "uri") { (uri: String) in
            try nativeAssertEquals(mockArgs?.path.absoluteString, uri)
        }
    }
    
    private func assertSetPageArgs(args: [String: Any]) throws {
        let mockHasRun = VoxeetSDK.shared.filePresentation.updateHasRun
        try ifKeyExists(arg: args, key: "hasRun") { (hasRun: Bool) in
            try nativeAssertEquals(mockHasRun, hasRun)
        }
        
        let mockArgs = VoxeetSDK.shared.filePresentation.updateArgs
        try ifKeyExists(arg: args, key: "page") { (page: Int) in
            try nativeAssertEquals(mockArgs, page)
        }
    }
    
    private func assertStartArgs(args: [String: Any]) throws {
        let mockHasRun = VoxeetSDK.shared.filePresentation.startHasRun
        try ifKeyExists(arg: args, key: "hasRun") { (hasRun: Bool) in
            try nativeAssertEquals(mockHasRun, hasRun)
        }
        
        let mockArgs = VoxeetSDK.shared.filePresentation.startArgs
        try ifKeyExists(arg: args, key: "id") { (id: String) in
            try nativeAssertEquals(mockArgs?.id, id)
        }
        
        try ifKeyExists(arg: args, key: "imageCount") { (imageCount: Int) in
            try nativeAssertEquals(mockArgs?.imageCount, imageCount)
        }
    }
    
    private func assertStopArgs(args: [String: Any]) throws {
        let mockHasRun = VoxeetSDK.shared.filePresentation.stopHasRun
        try ifKeyExists(arg: args, key: "hasRun") { (hasRun: Bool) in
            try nativeAssertEquals(mockHasRun, hasRun)
        }
    }
    
    private func setGetImageReturn(args: [String: Any]) throws {
        try ifKeyExists(arg: args, key: "url", closure: { (url: String) in
            VoxeetSDK.shared.filePresentation.imageReturn = URL(string: url)
        })
    }
    
    private func assertGetImageArgs(args: [String: Any]) throws {
        let mockHasRun = VoxeetSDK.shared.filePresentation.imageHasRun
        try ifKeyExists(arg: args, key: "hasRun") { (hasRun: Bool) in
            try nativeAssertEquals(mockHasRun, hasRun)
        }
        
        let mockArgs = VoxeetSDK.shared.filePresentation.imageArgs
        try ifKeyExists(arg: args, key: "page") { (page: Int) in
            try nativeAssertEquals(mockArgs, page)
        }
    }
    
    private func setGetThumbnailReturn(args: [String: Any]) throws {
        try ifKeyExists(arg: args, key: "url", closure: { (url: String) in
            VoxeetSDK.shared.filePresentation.thumbnailReturn = URL(string: url)
        })
    }
    
    private func assertGetThumbnailArgs(args: [String: Any]) throws {
        let mockHasRun = VoxeetSDK.shared.filePresentation.thumbnailHasRun
        try ifKeyExists(arg: args, key: "hasRun") { (hasRun: Bool) in
            try nativeAssertEquals(mockHasRun, hasRun)
        }
        
        let mockArgs = VoxeetSDK.shared.filePresentation.thumbnailArgs
        try ifKeyExists(arg: args, key: "page") { (page: Int) in
            try nativeAssertEquals(mockArgs, page)
        }
    }
}

extension FilePresentationServiceAsserts: SDKAsserts {
    
    public func runAssert(_ assert: String, args: [String: Any], result: (SDKAssertResult) -> Void) {
        do {
            switch assert {
                
            case "assertGetCurrentArgs":
                try assertGetCurrentArgs(args: args)
                
            case "assertConvertArgs":
                try assertConvertArgs(args: args)
                
            case "assertSetPageArgs":
                try assertSetPageArgs(args: args)
                
            case "assertStartArgs":
                try assertStartArgs(args: args)
                
            case "assertStopArgs":
                try assertStopArgs(args: args)
                
            case "setGetImageReturn":
                try setGetImageReturn(args: args)
                
            case "assertGetImageArgs":
                try assertGetImageArgs(args: args)
                
            case "setGetThumbnailReturn":
                try setGetThumbnailReturn(args: args)
                
            case "assertGetThumbnailArgs":
                try assertGetThumbnailArgs(args: args)
                
            default:
                throw SDKAssertError.unknownAssert
            }
            result(.success(()))
        } catch {
            result(SDKAssertError.toResult(error: error))
        }
    }
}

